state/presence/presence.go

```diff
@@ -757,6 +861,174 @@ func (p *Pinger) ping() (err error) {
 	return errors.Trace(err)
 }
 
+// collapsedBeingsInfo tracks the result of aggregating all of the items in the
+// beings table by their key.
+type collapsedBeingsInfo struct {
+	Key    string   `bson:"_id"`
+	Seqs    []int64 `bson:"seqs"`
+}
+
+
+// BeingPruner tracks the state of removing unworthy beings from the
+// presence.beings collection. Being sequences are unworthy once their sequence
+// has been superseded.
+type BeingPruner struct {
+	modelUUID string
+	beingsC *mgo.Collection
+	pingsC *mgo.Collection
+	toRemove []string
+	maxQueue int
+	removedCount uint64
+	delta time.Duration
+}
+
+// iterKeys is returns an iterator of Keys from this modelUUId and what Sequences
+// are in use to represent them.
+func (p *BeingPruner) iterKeys() *mgo.Iter {
+	thisModelRegex := bson.M{"_id": bson.M{"$regex": bson.RegEx{"^" + p.modelUUID, ""}}}
+	pipe := p.beingsC.Pipe([]bson.M{
+		// Grab all sequences for this model
+		{"$match": thisModelRegex},
+		// We don't need the _id
+		{"$project": bson.M{"_id": 0, "seq": 1, "key": 1}},
+		// Group all the sequences by their key.
+		{"$group": bson.M{
+			"_id":    "$key",
+			"seqs":    bson.M{"$push": "$seq"},
+		}},
+		// Filter out any keys that have only a single sequence
+		// representing them
+		// Note: indexing is from 0, you can set this to 2 if you wanted
+		// to only bother pruning sequences that have >2 entries.
+		// This mostly helps the 'nothing to do' case, dropping the time
+		// to realize there are no sequences to be removed from 36ms,
+		// down to 15ms with 3500 keys.
+		{"$match": bson.M{"seqs.1": bson.M{"$exists": 1}}},
+	})
+	pipe.Batch(1600)
+	return pipe.Iter()
+}
+
+// queueRemoval includes this sequence as one that has been superseded
+func (p *BeingPruner) queueRemoval(seq int64) {
+	p.toRemove = append(p.toRemove, docIDInt64(p.modelUUID, seq))
+}
+
+// flushRemovals makes sure that we've applied all desired removals
+func (p *BeingPruner) flushRemovals() error {
+	if len(p.toRemove) == 0 {
+		return nil
+	}
+	matched, err := p.beingsC.RemoveAll(bson.M{"_id": bson.M{"$in": p.toRemove}})
+	p.toRemove = p.toRemove[:0]
+	if matched.Removed > 0 {
+		p.removedCount += uint64(matched.Removed)
+	}
+	return err
+}
+
+func (p *BeingPruner) removeOldPings() error {
+	// now and now-period are both considered active slots, so we don't
+	// touch those. We also leave 2 more slots around
+	startTime := time.Now()
+	// TODO(jam): 2017-04-18 startCount may not be appropriate, because
+	// Count() includes all models, not just the one we are pruning
+	startCount, err := p.pingsC.Count()
```

Comment:
I think either do a count of ping for model (which should use the index so should be fairly fast) or just leave it out. A count across all models is confusing and may leave people wondering why so many are being left behind.